trie/stacktrie.go

```diff
@@ -0,0 +1,529 @@
+// Copyright 2020 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"io"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"golang.org/x/crypto/sha3"
+)
+
+// StackTrie is a trie implementation that expects keys to be inserted
+// in order. Once it determines that a subtree will no longer be inserted
+// into, it will hash it and free up the memory it uses.
+type StackTrie struct {
+	nodeType  uint8          // node type (as in branch, ext, leaf)
+	val       []byte         // value contained by this node if it's a leaf
+	key       []byte         // key chunk covered by this (full|ext) node
+	keyOffset int            // offset of the key chunk inside a full key
+	children  [16]*StackTrie // list of children (for fullnodes and exts)
+}
+
+// NewStackTrie allocates and initializes an empty trie.
+func NewStackTrie() *StackTrie {
+	return &StackTrie{
+		nodeType: emptyNode,
+	}
+}
+
+// List all values that StackTrie#nodeType can hold
+const (
+	emptyNode = iota
+	branchNode
+	extNode
+	leafNode
+	hashedNode
+)
+
+func (st *StackTrie) TryUpdate(key, value []byte) error {
+	k := keybytesToHex(key)
+	if len(value) == 0 {
+		panic("deletion not supported")
+	}
+	st.insert(k[:len(k)-1], value)
+	return nil
+}
+
+// Helper function that, given a full key, determines the index
+// at which the chunk pointed by st.keyOffset is different from
+// the same chunk in the full key.
+func (st *StackTrie) getDiffIndex(key []byte) int {
+	diffindex := 0
+	for ; diffindex < len(st.key) && st.key[diffindex] == key[st.keyOffset+diffindex]; diffindex++ {
+	}
+	return diffindex
+}
+
+// Helper function to that inserts a (key, value) pair into
+// the trie.
+func (st *StackTrie) insert(key, value []byte) {
+	switch st.nodeType {
+	case branchNode: /* Branch */
+		idx := int(key[st.keyOffset])
+		if st.children[idx] == nil {
+			st.children[idx] = NewStackTrie()
+			st.children[idx].keyOffset = st.keyOffset + 1
+		}
+		for i := idx - 1; i >= 0; i-- {
+			if st.children[i] != nil {
+				if st.children[i].nodeType != hashedNode {
+					st.children[i].val = st.children[i].hash()
+					st.children[i].key = nil
+					st.children[i].nodeType = hashedNode
```

Comment:
Vague idea here, what if there was a method like `intoHash`. Then you could write:
```
if st.children[i].nodeType != hashedNode{ 
   st.children[i].convertToHash()
}
```
And maybe that could be used later on too, maybe we can rewrite
```
	p.children[origIdx] = NewStackTrie()
		p.children[origIdx].nodeType = leafNode
		p.children[origIdx].key = st.key[diffidx+1:]
		p.children[origIdx].val = st.val
		p.children[origIdx].keyOffset = p.keyOffset + 1

		p.children[origIdx].val = p.children[origIdx].hash()
		p.children[origIdx].nodeType = hashedNode
``` 
into something like 
```
p.children[origIdx] = NewLeaf(st.key[diffidx+1:], val, p.keyOffset + 1).convertToHash()
```

Reply:
... and if we have something like `newLeaf`, then we could more easily experiment with e.g. having a temporary and non-synced pool of leafs which we reuse.